src/core/lib/event_engine/cf_engine/cftype_unique_ref.h

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GRPC_SRC_CORE_LIB_EVENT_ENGINE_CF_ENGINE_CFTYPE_UNIQUE_REF_H
+#define GRPC_SRC_CORE_LIB_EVENT_ENGINE_CF_ENGINE_CFTYPE_UNIQUE_REF_H
+#include <grpc/support/port_platform.h>
+
+#ifdef GPR_APPLE
+
+#include <CoreFoundation/CoreFoundation.h>
+
+namespace grpc_event_engine {
+namespace experimental {
+
+template <typename T>
+class CFTypeUniqueRef {
```

Comment:
@drfloob Hey AJ I wonder there are existing utilities like this in the grpc codebase?

This class seems quite generic to me so i'm hoping we could reuse if there exists one. Thanks!

Reply:
It looks like this class provides some sugar specifically for `CFTypeRef` objects (implicit conversions and output-parameter specializations). It looks to me like a `std::unique_ptr` could do this job, albeit without the sugar on top. I'm mostly ambivalent about it, but we don't have anything `CFTypeRef`-specific in the core codebase.

Reply:
This looks like `template <typename T> using CFTypeUniqueRef = std::unique_ptr<T, CFRelease>;`

Reply:
I tries with something like
```
template <typename T>
using CFTypeUniqueRef =
    std::unique_ptr<std::remove_cvref_t<decltype(*std::declval<T>())>,
                    decltype(& CFRelease)>;
```
Which is very difficult to use because:
1. Unlike typical unique_ptr use case, `T` (CFStringRef, CFErrorRef, etc.) is the pointer type thus the complex alias
2. Deleter (&CFRelease) needs to be passed everywhere creating a pointer
2. I don't do dereference (operator* and operator->) but only convert back and forth when using with CF APIs.
3. I need `operator&` for using as output parameter

Reply:
Alternative to this, I can also try warp a unique pointer for auto release then add implicit conversions and `operator&` if you prefer. It'll be something like
```
template <typename T>
class CFTypeUniqueRef {
public:
  // conversions, operator&
private:
  std::unique_ptr<std::remove_cvref_t<decltype(*std::declval<T>())>,
                    decltype(& CFRelease)> ref_;
};
```